modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3102,14 +3103,37 @@ void GDScriptAnalyzer::reduce_identifier(GDScriptParser::IdentifierNode *p_ident
 			GDScriptParser::DataType result;
 			result.kind = GDScriptParser::DataType::NATIVE;
 			result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-			if (autoload.path.to_lower().ends_with(GDScriptLanguage::get_singleton()->get_extension())) {
+			if (ResourceLoader::get_resource_type(autoload.path) == "GDScript") {
 				Ref<GDScriptParserRef> singl_parser = get_parser_for(autoload.path);
 				if (singl_parser.is_valid()) {
 					Error err = singl_parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED);
 					if (err == OK) {
 						result = type_from_metatype(singl_parser->get_parser()->head->get_datatype());
 					}
 				}
+			} else if (ResourceLoader::get_resource_type(autoload.path) == "PackedScene") {
+				Ref<PackedScene> scene = GDScriptCache::get_scene(autoload.path);
+				if (scene->get_state().is_valid()) {
+					const int ROOT_NODE = 0;
+					Ref<SceneState> state = scene->get_state();
+					for (int i = 0; i < state->get_node_property_count(ROOT_NODE); i++) {
```

Comment:
Why only check the root node's `script` property? How does this approach handle, for instance, scripts referenced by subnodes, potentially from custom properties?

Reply:
Because we're in `reduce_identifier`. Let's say that I have `MyAutoload.tscn` loaded as an autoload `MyAutoload`. That part checks if there's an autoload scene named `MyAutoload`. If so, it returns the main script.

But I need to test, as you said, subnodes and other instances.

Otherwise, maybe I could add support for generic scripts (with added checks for GDScript afterward).